File: ios/PTimerKit/Sources/PTimerCore/Timer/TimerState.swift
import Foundation

/// Numerical tolerance (1 microsecond) for the timer state machine's
/// wall-clock comparisons: deciding when a running timer has reached
/// its end date, and when a remaining-time value is effectively zero.
/// It absorbs floating-point drift at those boundaries so a timer
/// within 1 µs of its end is treated as completed rather than lingering
/// as `running`, and sub-microsecond remaining times snap to exactly 0.
///
/// This is the timer core's own constant. It is intentionally
/// independent of the exposure engine's `stabilityEpsilon`: the two
/// share the 1e-6 magnitude by coincidence but serve unrelated domains
/// (time vs. exposure-stop comparisons), so the timer core carries no
/// dependency on — and no sync obligation with — the exposure engine.
public let timerStabilityEpsilon: TimeInterval = 0.000_001

public enum TimerStatus: String, Equatable {
    case running
    // `paused` is a frozen, resumable state that preserves remaining time.
    case paused
    case completed
}

/// Payload of a `running` timer. Holds only the fields valid in the
/// `running` state: identity, duration, creation time, and the
/// expected end date.
public struct RunningTimer: Equatable {
    public let id: UUID
    public let duration: TimeInterval
    public let startDate: Date
    public let endDate: Date

    public init(id: UUID, duration: TimeInterval, startDate: Date, endDate: Date) {
        self.id = id
        self.duration = duration
        self.startDate = startDate
        self.endDate = endDate
    }
}

/// Payload of a `paused` timer. Holds only the freeze metadata
/// (`pausedRemainingTime` + `pausedAt`); the hypothetical completion
/// date is derived from `pausedAt + pausedRemainingTime` rather than
/// stored, matching Timer Spec §3.1 ("expectedCompletionAt — running
/// status only"). Mathematically equivalent to the previous stored
/// `endDate` because `pausing(at:)` always set `pausedRemainingTime
/// = endDate - pausedAt`.
public struct PausedTimer: Equatable {
    public let id: UUID
    public let duration: TimeInterval
    public let startDate: Date
    public let pausedRemainingTime: TimeInterval
    public let pausedAt: Date

    public init(id: UUID, duration: TimeInterval, startDate: Date, pausedRemainingTime: TimeInterval, pausedAt: Date) {
        self.id = id
        self.duration = duration
        self.startDate = startDate
        self.pausedRemainingTime = pausedRemainingTime
        self.pausedAt = pausedAt
    }

    /// Hypothetical completion date for display purposes. Derived from
    /// the freeze metadata so the field disappears from the persisted
    /// schema (Timer Spec §3.1) while UI consumers can still read it.
    public var endDate: Date {
        pausedAt.addingTimeInterval(pausedRemainingTime)
    }
}

/// Payload of a `completed` timer. Holds the recorded completion
/// timestamp; surfaced via `endDate` for backward-compatible callers
/// that read the legacy field.
public struct CompletedTimer: Equatable {
    public let id: UUID
    public let duration: TimeInterval
    public let startDate: Date
    public let completedAt: Date

    public init(id: UUID, duration: TimeInterval, startDate: Date, completedAt: Date) {
        self.id = id
        self.duration = duration
        self.startDate = startDate
        self.completedAt = completedAt
    }
}

/// Sum-type representation of a timer's lifecycle state. Each case
/// carries only the fields valid for that state, so invalid
/// combinations (e.g. running with a `pausedAt`) cannot be
/// constructed.
///
/// Backward-compatible computed properties (`endDate`,
/// `pausedRemainingTime`, `pausedAt`, `status`) preserve the legacy
/// struct surface so existing call sites and the persisted snapshot
/// schema stay byte-identical.
public enum TimerState: Identifiable, Equatable {
    case running(RunningTimer)
    case paused(PausedTimer)
    case completed(CompletedTimer)

    /// Compatibility initializer that mirrors the historical struct
    /// `TimerState(id:duration:startDate:endDate:pausedRemainingTime:pausedAt:status:)`
    /// constructor used by tests and the `PersistentTimerSnapshot`
    /// restore path. The initializer dispatches on `status` and
    /// constructs the appropriate sum case from the supplied legacy
    /// fields.
    ///
    /// Trusted-callsite contract: callers shall supply the fields
    /// required by the chosen status:
    /// - `.running` ⇒ `endDate` non-nil
    /// - `.paused` ⇒ `pausedRemainingTime` and `pausedAt` non-nil
    /// - `.completed` ⇒ `endDate` non-nil (= completion timestamp)
    /// `endDate` is intentionally ignored for `.paused`; the sum-type
    /// representation derives it from `pausedAt + pausedRemainingTime`
    /// (Timer Spec §3.1 "expectedCompletionAt — running status only").
    /// Other missing fields debug-trap so corrupt inputs from a
    /// persisted snapshot are caught early; the production fallback
    /// constructs a degenerate but type-valid case so a debug crash
    /// does not become a release crash. `PersistentTimerSnapshot.restore`
    /// guards the `.paused` corrupt-input case at its caller boundary
    /// instead, surfacing such snapshots as completed.
    public init(
        id: UUID,
        duration: TimeInterval,
        startDate: Date,
        endDate: Date?,
        pausedRemainingTime: TimeInterval?,
        pausedAt: Date?,
        status: TimerStatus
    ) {
        switch status {
        case .running:
            assert(endDate != nil, "TimerState(.running) requires a non-nil endDate")
            self = .running(
                RunningTimer(
                    id: id,
                    duration: duration,
                    startDate: startDate,
                    endDate: endDate ?? startDate.addingTimeInterval(duration)
                )
            )
        case .paused:
            assert(
                pausedRemainingTime != nil && pausedAt != nil,
                "TimerState(.paused) requires non-nil pausedRemainingTime and pausedAt"
            )
            self = .paused(
                PausedTimer(
                    id: id,
                    duration: duration,
                    startDate: startDate,
                    pausedRemainingTime: pausedRemainingTime ?? 0,
                    pausedAt: pausedAt ?? startDate
                )
            )
        case .completed:
            assert(endDate != nil, "TimerState(.completed) requires a non-nil endDate (completion timestamp)")
            self = .completed(
                CompletedTimer(
                    id: id,
                    duration: duration,
                    startDate: startDate,
                    completedAt: endDate ?? startDate.addingTimeInterval(duration)
                )
            )
        }
    }

    public var id: UUID {
        switch self {
        case .running(let payload): return payload.id
        case .paused(let payload): return payload.id
        case .completed(let payload): return payload.id
        }
    }

    public var duration: TimeInterval {
        switch self {
        case .running(let payload): return payload.duration
        case .paused(let payload): return payload.duration
        case .completed(let payload): return payload.duration
        }
    }

    public var startDate: Date {
        switch self {
        case .running(let payload): return payload.startDate
        case .paused(let payload): return payload.startDate
        case .completed(let payload): return payload.startDate
        }
    }

    /// Backward-compatible field surface. `endDate` is non-nil for
    /// every case in the sum-type representation; the optional return
    /// type is preserved so callers reading the legacy field continue
    /// to compile and behave identically.
    public var endDate: Date? {
        switch self {
        case .running(let payload): return payload.endDate
        case .paused(let payload): return payload.endDate
        case .completed(let payload): return payload.completedAt
        }
    }

    public var pausedRemainingTime: TimeInterval? {
        if case .paused(let payload) = self {
            return payload.pausedRemainingTime
        }
        return nil
    }

    public var pausedAt: Date? {
        if case .paused(let payload) = self {
            return payload.pausedAt
        }
        return nil
    }

    /// Derived status accessor preserving the legacy `status`
    /// property. External callers (display state mappers, lock-
    /// screen coordinator, view models) keep reading this without
    /// switching on case.
    public var status: TimerStatus {
        switch self {
        case .running: return .running
        case .paused: return .paused
        case .completed: return .completed
        }
    }

    public var remainingTime: TimeInterval {
        assert(duration.isFinite && duration > 0, "Timer duration must be finite and positive.")
        switch self {
        case .running(let payload):
            return Self.sanitizeRemainingTime(payload.endDate.timeIntervalSinceNow)
        case .paused(let payload):
            return Self.sanitizeRemainingTime(payload.pausedRemainingTime)
        case .completed:
            return 0
        }
    }

    public func remainingTime(at now: Date) -> TimeInterval {
        assert(duration.isFinite && duration > 0, "Timer duration must be finite and positive.")
        switch self {
        case .running(let payload):
            return Self.sanitizeRemainingTime(payload.endDate.timeIntervalSince(now))
        case .paused(let payload):
            return Self.sanitizeRemainingTime(payload.pausedRemainingTime)
        case .completed:
            return 0
        }
    }

    public func status(at now: Date) -> TimerStatus {
        guard case .running(let payload) = self,
              now.addingTimeInterval(timerStabilityEpsilon) >= payload.endDate else {
            return status
        }

        return .completed
    }

    public func updatingStatus(at now: Date) -> TimerState {
        guard case .running(let payload) = self,
              now.addingTimeInterval(timerStabilityEpsilon) >= payload.endDate else {
            return self
        }

        return completed(at: payload.endDate)
    }

    public func pausing(at now: Date) -> TimerState {
        let remaining = remainingTime(at: now)

        guard remaining > 0 else {
            // Preserve the legacy short-circuit: pausing while remaining
            // time has reached zero immediately completes using the
            // existing endDate (or `now` as the fallback only if endDate
            // is somehow absent — unreachable in the sum-type
            // representation but kept for shape parity).
            return completed(at: endDate ?? now)
        }

        // Freeze the timer with its remaining duration intact so it can be
        // resumed later from the same logical point. The hypothetical
        // completion date stays accessible via the computed
        // `PausedTimer.endDate` (= `pausedAt + pausedRemainingTime`),
        // matching the legacy stored value exactly.
        return .paused(
            PausedTimer(
                id: id,
                duration: duration,
                startDate: startDate,
                pausedRemainingTime: remaining,
                pausedAt: now
            )
        )
    }

    public func resume(at now: Date) -> TimerState {
        let remaining = Self.sanitizeRemainingTime(pausedRemainingTime ?? 0)
        guard remaining > 0 else {
            return completed(at: resolvedCompletionDate())
        }

        // Resume recalculates the end date from "now" because `paused`
        // preserves remaining time as a frozen resumable state.
        return .running(
            RunningTimer(
                id: id,
                duration: duration,
                startDate: startDate,
                endDate: now.addingTimeInterval(remaining)
            )
        )
    }

    public func completed(at completionDate: Date? = nil) -> TimerState {
        .completed(
            CompletedTimer(
                id: id,
                duration: duration,
                startDate: startDate,
                completedAt: completionDate ?? resolvedCompletionDate()
            )
        )
    }

    private func resolvedCompletionDate() -> Date {
        // Mirrors the legacy `endDate ?? pausedAt + remaining ??
        // startDate + duration` resolution exactly. In the sum-type
        // representation `endDate` is non-nil for running and paused
        // and `completedAt` is the recorded completion timestamp,
        // so the legacy fallback chain collapses to a direct case
        // dispatch without changing observable behavior.
        switch self {
        case .running(let payload):
            return payload.endDate
        case .paused(let payload):
            return payload.endDate
        case .completed(let payload):
            return payload.completedAt
        }
    }

    private static func sanitizeRemainingTime(_ value: TimeInterval) -> TimeInterval {
        let clamped = max(0, value)
        return clamped < timerStabilityEpsilon ? 0 : clamped
    }
}

File: android/core/src/main/kotlin/com/sangwook/ptimer/core/timer/TimerState.kt
package com.sangwook.ptimer.core.timer

import java.time.Duration
import java.time.Instant
import java.util.UUID
import kotlin.math.max
import kotlin.math.roundToLong

// Faithful port of iOS PTimerCore TimerState (PROTECTED AREA — the pause/
// resume/complete/cancel state machine). Dates map to java.time.Instant and
// TimeInterval to Double seconds; comparisons use nanosecond precision so the
// 1 µs stability epsilon behaves as on iOS.

/** 1 microsecond tolerance for the timer state machine's wall-clock comparisons. */
const val TIMER_STABILITY_EPSILON: Double = 0.000_001

internal fun secondsBetween(from: Instant, to: Instant): Double =
    Duration.between(from, to).toNanos() / 1_000_000_000.0

internal fun Instant.plusSecondsDouble(seconds: Double): Instant =
    this.plusNanos((seconds * 1_000_000_000.0).roundToLong())

enum class TimerStatus { running, paused, completed, canceled }

/**
 * Sum-type representation of a timer's lifecycle. Each case carries only the
 * fields valid for that state, so invalid combinations cannot be constructed.
 */
sealed interface TimerState {
    val id: UUID
    val duration: Double
    val startDate: Instant

    data class Running(
        override val id: UUID,
        override val duration: Double,
        override val startDate: Instant,
        val endDate: Instant,
    ) : TimerState

    data class Paused(
        override val id: UUID,
        override val duration: Double,
        override val startDate: Instant,
        val pausedRemainingTime: Double,
        val pausedAt: Instant,
    ) : TimerState {
        /** Hypothetical completion date, derived (not stored). */
        val endDate: Instant get() = pausedAt.plusSecondsDouble(pausedRemainingTime)
    }

    data class Completed(
        override val id: UUID,
        override val duration: Double,
        override val startDate: Instant,
        val completedAt: Instant,
    ) : TimerState

    data class Canceled(
        override val id: UUID,
        override val duration: Double,
        override val startDate: Instant,
        val canceledAt: Instant,
        /** Remaining time captured at cancellation (e.g. "Canceled · 51s left"). */
        val remainingAtCancel: Double,
    ) : TimerState

    companion object {
        /**
         * Compatibility factory mirroring the legacy field-based constructor
         * used by the persisted-snapshot restore path. Trusted-callsite
         * contract: supply the fields required by [status]. `endDate` is
         * ignored for `.paused` (derived from `pausedAt + pausedRemainingTime`).
         */
        fun fromLegacy(
            id: UUID,
            duration: Double,
            startDate: Instant,
            endDate: Instant?,
            pausedRemainingTime: Double?,
            pausedAt: Instant?,
            status: TimerStatus,
        ): TimerState = when (status) {
            TimerStatus.running -> Running(
                id, duration, startDate,
                endDate = endDate ?: startDate.plusSecondsDouble(duration),
            )
            TimerStatus.paused -> Paused(
                id, duration, startDate,
                pausedRemainingTime = pausedRemainingTime ?: 0.0,
                pausedAt = pausedAt ?: startDate,
            )
            TimerStatus.completed -> Completed(
                id, duration, startDate,
                completedAt = endDate ?: startDate.plusSecondsDouble(duration),
            )
            TimerStatus.canceled -> Canceled(
                id, duration, startDate,
                canceledAt = endDate ?: startDate.plusSecondsDouble(duration),
                remainingAtCancel = pausedRemainingTime ?: 0.0,
            )
        }
    }
}

val TimerState.status: TimerStatus
    get() = when (this) {
        is TimerState.Running -> TimerStatus.running
        is TimerState.Paused -> TimerStatus.paused
        is TimerState.Completed -> TimerStatus.completed
        is TimerState.Canceled -> TimerStatus.canceled
    }

/** Non-null in every case (running/paused expected completion; terminal stamp). */
val TimerState.endDate: Instant
    get() = when (this) {
        is TimerState.Running -> endDate
        is TimerState.Paused -> endDate
        is TimerState.Completed -> completedAt
        is TimerState.Canceled -> canceledAt
    }

val TimerState.pausedRemainingTime: Double?
    get() = (this as? TimerState.Paused)?.pausedRemainingTime

val TimerState.pausedAt: Instant?
    get() = (this as? TimerState.Paused)?.pausedAt

val TimerState.remainingAtCancel: Double?
    get() = (this as? TimerState.Canceled)?.remainingAtCancel

fun TimerState.remainingTime(at: Instant): Double = when (this) {
    is TimerState.Running -> sanitizeRemainingTime(secondsBetween(at, endDate))
    is TimerState.Paused -> sanitizeRemainingTime(pausedRemainingTime)
    is TimerState.Completed, is TimerState.Canceled -> 0.0
}

fun TimerState.status(at: Instant): TimerStatus {
    if (this is TimerState.Running && secondsBetween(at, endDate) <= TIMER_STABILITY_EPSILON) {
        return TimerStatus.completed
    }
    return status
}

fun TimerState.updatingStatus(at: Instant): TimerState {
    if (this is TimerState.Running && secondsBetween(at, endDate) <= TIMER_STABILITY_EPSILON) {
        return completed(at = endDate)
    }
    return this
}

fun TimerState.pausing(at: Instant): TimerState {
    val remaining = remainingTime(at)
    if (remaining <= 0) {
        return completed(at = endDate)
    }
    return TimerState.Paused(
        id = id,
        duration = duration,
        startDate = startDate,
        pausedRemainingTime = remaining,
        pausedAt = at,
    )
}

fun TimerState.resume(at: Instant): TimerState {
    val remaining = sanitizeRemainingTime(pausedRemainingTime ?: 0.0)
    if (remaining <= 0) {
        return completed(at = resolvedCompletionDate())
    }
    return TimerState.Running(
        id = id,
        duration = duration,
        startDate = startDate,
        endDate = at.plusSecondsDouble(remaining),
    )
}

fun TimerState.completed(at: Instant? = null): TimerState = TimerState.Completed(
    id = id,
    duration = duration,
    startDate = startDate,
    completedAt = at ?: resolvedCompletionDate(),
)

/**
 * Transitions a running or paused timer to the terminal canceled record.
 * Already-terminal states are returned unchanged.
 */
fun TimerState.canceled(at: Instant): TimerState = when (this) {
    is TimerState.Running, is TimerState.Paused -> TimerState.Canceled(
        id = id,
        duration = duration,
        startDate = startDate,
        canceledAt = at,
        remainingAtCancel = remainingTime(at),
    )
    is TimerState.Completed, is TimerState.Canceled -> this
}

private fun TimerState.resolvedCompletionDate(): Instant = when (this) {
    is TimerState.Running -> endDate
    is TimerState.Paused -> endDate
    is TimerState.Completed -> completedAt
    is TimerState.Canceled -> canceledAt
}

private fun sanitizeRemainingTime(value: Double): Double {
    val clamped = max(0.0, value)
    return if (clamped < TIMER_STABILITY_EPSILON) 0.0 else clamped
}
